python/katana/local/_graph.pyx

```diff
@@ -229,15 +229,16 @@ cdef class GraphBase:
         """
         handle_result_void(self.underlying_property_graph().AddNodeProperties(GraphBase._convert_table(table, kwargs)))
 
-    def upsert_node_property(self, table=None, **kwargs):
+    def upsert_node_property(self, TxnContext ctx=TxnContext(), table=None, **kwargs):
```

Comment:
This needs to be done for all the functions you changed in this file.

Reply:
Oh a better example of why this change is required is `upsert_node_property(user_dataframe)`.

Reply:
Isn't there some style in Python where the default argument values should be `None` or pure values as otherwise `ctx` in `ctx=TxnContext()` will be shared for all invocations and a mutation to `ctx` will leak to other invocations?

Reply:
Oh yes!! Good catch. I forgot that. You must have None as the default and then you can use `ctx or TxnContext()` inside to create a new context iff the argument isn't already a real context.